naivedatetimefield/__init__.py

```diff
@@ -112,104 +122,138 @@ def to_python(self, value):
         )
 
     def get_prep_value(self, value):
-        """
-        Ensure we have a naive datetime ready for insertion
-        """
-        value = super(NaiveDateTimeField, self).get_prep_value(value)
-        value = self.to_python(value)
-
-        if value is not None and timezone.is_aware(value):
-            # We were given an aware datetime, strip off tzinfo
-            value = value.replace(tzinfo=None)
-
-        return value
-
-    def get_db_prep_value(self, value, connection, prepared=False):
-        if not prepared:
-            value = self.get_prep_value(value)
-
-        if value is None:
-            return None
-
-        if hasattr(value, "resolve_expression"):
+        return super(DateTimeField, self).get_prep_value(value)
+
+    def from_db_value(self, value, expression, connection):
+        is_truncbase = isinstance(expression, TruncBase)
+        if is_truncbase and not isinstance(expression, NaiveAsSQLMixin):
+            raise TypeError(
+                "Django's %s cannot be used with a NaiveDateTimeField"
+                % expression.__class__.__name__
+            )
+        if connection.vendor == "postgresql":
+            if is_truncbase:
+                return timezone.make_naive(value, pytz.utc)
             return value
-
-        if connection.settings_dict["ENGINE"] == "django.db.backends.mysql":
-            return str(value)
-
-        elif connection.settings_dict["ENGINE"] == "django.db.backends.sqlite3":
-            return str(value)
-
-        elif connection.settings_dict["ENGINE"] == "django.db.backends.oracle":
-            from django.db.backends.oracle.utils import Oracle_datetime
-
-            return Oracle_datetime.from_datetime(value)
-
+        if timezone.is_aware(value):
+            return timezone.make_naive(value, connection.timezone)
         return value
 
     def pre_save(self, model_instance, add):
         if self.auto_now or (self.auto_now_add and add):
-            value = timezone.now().replace(tzinfo=None)
+            value = timezone.make_naive(timezone.now())
             setattr(model_instance, self.attname, value)
             return value
         else:
             return super(NaiveDateTimeField, self).pre_save(model_instance, add)
 
-    def value_to_string(self, obj):
-        val = self.value_from_object(obj)
-        return "" if val is None else val.isoformat()
-
-    def formfield(self, **kwargs):
-        defaults = {"form_class": forms.DateTimeField}
-        defaults.update(kwargs)
-        return super(NaiveDateTimeField, self).formfield(**defaults)
 
+class NaiveTimezoneMixin(object):
+    def get_tzname(self):
+        if isinstance(self.output_field, NaiveDateTimeField):
+            if self.tzinfo is not None:
+                warnings.warn(
+                    "tzinfo argument provided when truncating a NaiveDateTimeField. "
+                    "This argument will have no effect."
+                )
+            return "UTC"
```

Comment:
Is there a constant we can use here instead?

Reply:
Yes, good call. This is part of a pretty convoluted process to lead the DB backend to produce the right result. I'll look at making this a constant and commenting more thoroughly.

Reply:
OK, I started looking deeper into why various things work and it led me down a rabbit hole. I noticed that if you use the `TIME_ZONE` setting in your database settings dict things break. So we now support that, and the implementation is overall a lot clearer IMO.

Reply:
Whoa, nice find! 👍